### ISSUE TITLE
[MNG-3057] properties not expanded in generated POMs when building A/B/C nested projects

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#4843]
**[George Armhold](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=armhold)** opened **[MNG-3057](https://issues.apache.org/jira/browse/MNG-3057?redirect=false)** and commented

Using Maven version: 2.0.8-SNAPSHOT, svn r547427.

I checked out and built 2.0.8 because I was interested in Jason van Zyl's patch for MNG-2619- "building from the middle pom of a (parent,child,grandchild) heirarchy fails".  The fix works for the most part, but there still seems to be a problem with the poms that get generated during the install goal.  The poms that get installed into .m2/repository do not have properties interpolated.  If I run maven like:

$ mvn install -Dglobal-version=1.0.0

I get poms with the string "${global-version}" embedded in the resulting poms rather than what I specify on the command line (or in settings.xml).  The build works properly aside from this, and if I do "mvn help:effective-pom" the properties are correctly interpolated.

I've attached a tarfile with an example A/B/C build structure that exhibits this behavior, as well as the generated poms.

Many thanks for your attention.

---

**Affects:** 2.0.7

**Attachments:**
- [example.tar.gz](https://issues.apache.org/jira/secure/attachment/12712448/example.tar.gz) (_5.75 kB_)
- [generated-poms.tar.gz](https://issues.apache.org/jira/secure/attachment/12712447/generated-poms.tar.gz) (_2.04 kB_)
- [InstallMojo.java.patch](https://issues.apache.org/jira/secure/attachment/12712262/InstallMojo.java.patch) (_3.50 kB_)
- [InstallMojo.java.patch](https://issues.apache.org/jira/secure/attachment/12712242/InstallMojo.java.patch) (_2.91 kB_)
- [InstallMojo.quoteReplacement.patch](https://issues.apache.org/jira/secure/attachment/12712798/InstallMojo.quoteReplacement.patch) (_572 bytes_)
- [maven-dependency-problem.zip](https://issues.apache.org/jira/secure/attachment/12712837/maven-dependency-problem.zip) (_2.29 kB_)
- [maven-install-parent.patch](https://issues.apache.org/jira/secure/attachment/12712267/maven-install-parent.patch) (_8.09 kB_)

**Issue Links:**
- [MDEPLOY-94](https://issues.apache.org/jira/browse/MDEPLOY-94) Variables in the version element in pom.xml need to be expanded before deployment
 (_**"is duplicated by"**_)
- [MINSTALL-50](https://issues.apache.org/jira/browse/MINSTALL-50) provide property filtering on .pom files placed in local repo
 (_**"is duplicated by"**_)
- [MNG-2971](https://issues.apache.org/jira/browse/MNG-2971) Variables are not replaced into installed pom file

- [MNG-4140](https://issues.apache.org/jira/browse/MNG-4140) Properties incorrectly replaced in pom

- [MNG-2412](https://issues.apache.org/jira/browse/MNG-2412) global variable filtering of pom.xml for parent and sub module pom.xml files is not working when deploying to a repository.

- [MNG-4223](https://issues.apache.org/jira/browse/MNG-4223) Resolve expressions in POM artifact coordinates

- [MNG-2446](https://issues.apache.org/jira/browse/MNG-2446) parent Pom  properties not resolved for module dependencies

- [MNG-624](https://issues.apache.org/jira/browse/MNG-624) automatic parent versioning

- [MNG-2619](https://issues.apache.org/jira/browse/MNG-2619) building from the middle pom of a (parent,child,grandchild) heirarchy fails

- [MNG-2820](https://issues.apache.org/jira/browse/MNG-2820) Deployment is stripping out the license header from the POM

- [MNG-3538](https://issues.apache.org/jira/browse/MNG-3538) Properties inside \<dependency> are not replaced in resulting POM

- [MNG-3782](https://issues.apache.org/jira/browse/MNG-3782) Variable substition not performed in transitive dependency using value from active profile

- [MNG-4167](https://issues.apache.org/jira/browse/MNG-4167) version-expression transformation interferes with plugins like GPG

9 votes, 7 watchers

**Comment by jira-importer** [comment]
**[Henrik Brautaset Aronsen](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=henrik242)** commented

I added another test case for this problem in MNG-2446.

**Comment by jira-importer** [comment]
**[Harsha Godugu](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hg2008)** commented

Hello:

The above patch does not work, if the global-version property is defined outside pom.xml or  outside settings.xml .  In other words, if you define like :

$ mvn install -Dglobal-version=1.0.0

in maven 2.0.9  it does not reflect the  property "global-version"  into the project properties (another existing bug against maven).
Hence,  the generated pom still will have stale ${global-version}  as value.

As you can guess  obviously,  one need to add a few  lines of code to fix the bug.  Check property against  System.getProperty(key) to  see if there is a global property defined from command line...

Here is a partial diff:

+ 

      while (m.find()) {

+ 

      String match = m.group();

+ 

      String key = match.substring(2, match.length()-1);

+ 

      String value = properties.getProperty(key);

+ 

      if (value == null) {

+ 

      System.out.println("Trying .. system wide property..");

+ 

      value = System.getProperty(key);

+ 

      if (value !=null) 

+ 

      System.out.println("Found value from sys=" + value);

+ 

      }

+ 

      if (value != null) {

+ 

      s = s.replaceAll("\\$\{" + key + "\}", value);

+ 

      }

**Comment by jira-importer** [comment]
**[Henrik Brautaset Aronsen](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=henrik242)** commented

I've updated [the patch|^InstallMojo.java.patch].  System properties are taken into account as well now.

**Comment by jira-importer** [comment]
**[Maxim Petrashev](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mpetrashev)** commented

Hi, Henrik

Can you, please,
- add supporting of nested values? I mean if we have property:
  someProperty1=/somepath/somesubpath/${otherPropery2}/something
  then it will fail with
  java.lang.IllegalArgumentException: Illegal group reference
  at java.util.regex.Matcher.appendReplacement(Matcher.java:706)
  at java.util.regex.Matcher.replaceAll(Matcher.java:806)
  at java.lang.String.replaceAll(String.java:2000)
  at org.apache.maven.plugin.install.InstallMojo.insertProperties(InstallMojo.java:118)
  where my InstallMojo.java:118 line is:
  s = s.replaceAll("\\$\{" + key + "\}", value);
- extend mask for "." sign in  property names? In you case {myComponent-version} works fine, but {myComponent.version} doesn't work
- I see just 1 place for properties replacement:
  if ( isPomArtifact )
  {
  installer.install( insertProperties(pomFile), artifact, localRepository );
  }
  how about else statement?
  else
  {
  metadata = new ProjectArtifactMetadata( artifact, pomFile );
  I believe we have to change it on
  metadata = new ProjectArtifactMetadata( artifact, insertProperties(pomFile) );
  too?
- Am I right that we need to extend deploy plugin too to make it configurable for pomFile property? Otherwise, original pom.xml will be deployed, not target/maven-install-plugin/pom.xml

Many thanks for your efforts! Does maven team have any plans to include this patch in the next release?

**Comment by jira-importer** [comment]
**[Henrik Brautaset Aronsen](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=henrik242)** commented

Hi Maxim.

I'm attaching [^maven-install-parent.patch], which is the one I've been using locally for awhile.  It fixes dots in property names and the properties replacements you mentioned, as well as the failing unit tests.

I am not using nested properties and mvn deploy myself, and haven't given that much thought -- sorry about that.

I don't think the maven team have any plans of including **this** patch, but there's ongoing work in MNG-624 to do something simliar which [might be included in Maven 2.1.0-M6](http://docs.codehaus.org/display/MAVEN/Maven+2.1.0+Release+Plan) .

**Comment by jira-importer** [comment]
**[John Dennis Casey](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jdcasey)** commented

moving back into 2.1.0 bucket for discussions related to MDEPLOY-94, etc. Maybe we should take a shot at this and start climbing the learning curve now in terms of use cases.

**Comment by jira-importer** [comment]
**[John Dennis Casey](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jdcasey)** commented

POM version-expression transformer is removing comments, because it's using the XPP3 reader/writer. I'm not sure how to work around this yet.

**Comment by jira-importer** [comment]
**[Benjamin Bentmann](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=bentmann)** commented

John, the removal of the license header was a regression of MNG-2820 for which I just added a (disabled) IT.

**Comment by jira-importer** [comment]
**[John Dennis Casey](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jdcasey)** commented

Added new interpolation strategy that uses raw contents of the POM XML file, and interpolates it from that string. This is a two-stage process, where the first stage interpolator isolates the version elements and hands the value off to the second-stage interpolator to actually resolve any expressions. Then, the result is re-inserted into the XML.

Also, adjusted the handling for artifacts with type == pom during install/deploy, such that the accompanying file is forcibly set to the artifact's file, to work around MCOMPILER-94. Attachments should be handled separately, since they have their own Artifact instance. I've modified the integration test for MNG-3057 to inject the configuration from MCOMPILER-94 just to be sure it works.

**Comment by jira-importer** [comment]
**[Henrik Brautaset Aronsen](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=henrik242)** commented

This issue is still not fixed in 2.1.0.

Steps to reproduce:
- download and unzip [^maven-dependency-problem.zip]
- cd first
- mvn clean install
- cd fourth
- mvn clean install  (which fails)

The project layout is this:

```
first--+--second----third
       |
       +--fourth
```

"third" has "second" as parent version, and "second"/"fourth" has "first" as parent version.  They all use a common applicationVersion property, which is set in "first".  "fourth" has a dependency on "third", which resolves OK, but it fails to resolve the transitive dependency to "second":

```
[INFO] Failed to resolve artifact.

GroupId: me
ArtifactId: second
Version: ${applicationVersion}

Reason: Unable to download the artifact from any repository

  me:second:pom:${applicationVersion}
```

**Comment by jira-importer** [comment]
**[Benjamin Bentmann](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=bentmann)** commented

The code that was introduced for this fix was removed from Maven 2.2+ due to reasons outlined in MNG-4223.
